packages/dds/counter/api-report/counter.api.md

```diff
@@ -21,7 +22,7 @@ export interface ISharedCounterEvents extends ISharedObjectEvents {
 }
 
 // @alpha
-export const SharedCounter: ISharedObjectKind<ISharedCounter>;
+export const SharedCounter: ISharedObjectKind<ISharedCounter> & SharedObjectKind<ISharedCounter>;
```

Comment:
Is it true that every place where we use ISharedObjectKind we will overlap with SharedObjectKind? If yes, should we reconsider how ISharedObjectKind is defined?

Reply:
If it were practical to do so yes. Sadly I don't think there is a good way to do it.

When dealing with type erasure and unimplementable interfaces, this gets a bit nasty.

If we make ISharedObjectKind extend SharedObjectKind, then its impossible to get typescript to type check an implementation of ISharedObjectKind since SharedObjectKind goes out of its way to opt out of structural typing and not be implementable (which it has to, or all values would be usable as ISharedObjectKinds).

If someone is using the encapsulated API, them manually implementing ISharedObjectKind is probably the main reason they would use the type, and it works fine with how the code is now. You only hit this problem if you write code which needs to work with both the encapsulated API, and also the declarative API. Currently I expect only our repo to do that, and it deduplicates all the type nastiness in one function (createSharedObjectKind): this type is implicitly inferred in the implementation as the return value from that function.

Reply:
I have updated some of the relevtn doc comments to help clarify this.